### PR TITLE
The resin maw : an end game FOB crusher xeno structure

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

This is a new xeno structure, the resin maw. Queen or shrike are able to morph one of their silo (small or regular) into a burrower tunnel. This will destroy all silo on the map and reset the psy points reserve of the hive.

Queen or shrike will be able to enter theresin maw, and chose an exit point anywhere on the map. This special tunnel will then burrow its way toward this point, creating after 5 minutes a tunnel between these two locations. An alert is gave to everyone when the resin maw starts digging.

When the head of themaw emerge (probably in the middle of marines fortification), it will delete everything on its 7x7 zone of arrival, and will stun every human nearby (10 tiles of range) for 5 seconds. Every xenos that were waiting in the other part of the worm will get expulsed from the worm.

All silos being dead, the only thing preventing siloless collapse will be both side of the maw If you manage to kill both parts, the timer begin as usual. 

Both side of the maw will be trackeable by pinpointers

## Why It's Good For The Game

This will be a solution for xeno to be able to breach huge fortification that are often here around FOB. This is not without risk though, since you are losing all your larva facilities. If marines manage to hold out that attack, they might be able to eradicate the whole hive

Known issues:

Unga will triple/quadruple cade inside the FOB. That will probably make the FOB harder to defend on the inside, but a possible solution is just to make the worm surface of destruction bigger than 3x3

Unga will OB the worm instantly, whatever the cost. Even if all xeno escaped, this would be a major drawback for the worm. Possible solution: make the worm fire/explosion immune. The only way to kill it would be to melee it

## Changelog
:cl:
add: The resin maw. A xeno structure able to spawn an exit point tunnel anywhere on the map, after 5 minutes of burrowing. You can create it by converting one of your silo into that special structure, but be aware that it will destroy all other silos and reset your psy points stock. 
/:cl:

### CODE REVIEW

**Comment by SpaceLoveSs13** [comment]
Now this is something good

**Comment by Swagile** [comment]
should up the rate from 3x3 to 6x6, or else marines just build cades around the xenos likely place of entry and own zone them

**Comment by gazzmaz4** [comment]
Not a crab 7/10

**Comment by XSlayer300** [comment]
the alaskan bull worm

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 1]
```diff
@@ -1050,6 +1050,44 @@
 	xeno_message("[X.name] has built a silo at [get_area(A)]!", "xenoannounce", 5, X.hivenumber)
 
 
+////////////////////
+/// Build giant worm
+///////////////////
+/datum/action/xeno_action/activable/build_giant_worm
+	name = "Transform to giant worm"
+	action_icon_state = "resin_silo"
+	mechanics_text = "Transform a silo into a giant burrower worm. Will destroy all our silo and reset our psych points reserve"
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:1059]:
```suggestion
	mechanics_text = "Transform a silo into a giant burrower worm. This ability will destroy all our silos in the field and reset our psych points reserve, however!"
```

File: code/modules/xenomorph/giant_worm.dm [hunk 2]
```diff
@@ -0,0 +1,242 @@
+/obj/structure/resin/giant_worm
+	name = "giant worm"
+	desc = "A giant burrowing worm created with the remains of a silo" //Need better name and desc
```
Comment (XSlayer300) [inline code/modules/xenomorph/giant_worm.dm:3]:
```suggestion
	desc = "A giant burrowing worm created with the remains of a silo. The mouth is agape and it looks alive." //Need better name and desc
```

File: code/modules/xenomorph/giant_worm.dm [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	. = ..()
+	GLOB.xeno_giant_worms -= src
+	associated_hive.handle_silo_death_timer()
+	for(var/mob/living/carbon/xenomorph/xeno AS in src)
+		xeno.forceMove(center_turf)
+	xeno_message("The [tail? "tail" :"head"] of the giant worm has been destroyed!", "xenoannounce", hivenumber = associated_hive.hivenumber)
+
+///Digging is starting, we notify everyone and kill all silos
+/obj/structure/resin/giant_worm/proc/start_digging(turf/target)
+	message_admins("A giant worm started digging at [AREACOORD(src)]")
```
Comment (XSlayer300) [inline code/modules/xenomorph/giant_worm.dm:52]:
```suggestion
	message_admins("A giant worm started digging at [AREACOORD(src)]!")
```

File: code/modules/xenomorph/giant_worm.dm [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	priority_announce("Sismic signal detected. An unknow object is burrowing under the planet surface") //Could use better name
```
Comment (XSlayer300) [inline code/modules/xenomorph/giant_worm.dm:53]:
```suggestion
	priority_announce("Warning: unusual seismic readings detected. Our data suggests that an unidentified entity is starting to burrow under the area of operations, likely affiliated by Xenomorphs. Terminate it as soon as possible.", title = "TGMC Intel Division")
```

File: code/modules/xenomorph/giant_worm.dm [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	xeno_message("A giant worm has started digging and will reach its destination in 5 minutes", "xenoannounce", hivenumber = associated_hive.hivenumber, target = src)
```
Comment (XSlayer300) [inline code/modules/xenomorph/giant_worm.dm:54]:
```suggestion
	xeno_message("A giant worm has started digging and will reach its destination in 5 minutes.", "xenoannounce", hivenumber = associated_hive.hivenumber, target = src)
```
I will get a better one soon.

File: code/modules/xenomorph/giant_worm.dm [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	name = "Start digging"
+	background_icon_state = "template2"
+	action_icon_state = "camera_off" //Need icon
+
+/datum/action/innate/select_exit_location/Activate()
+	var/mob/living/C = target
+	var/mob/camera/aiEye/remote/burrower_camera/eye = C.remote_control
+	var/obj/structure/resin/giant_worm/worm = eye.origin
+	if(!worm.checkExitSpot())
+		to_chat(target, "<span class='warning'>The worm head will not fit here")
```
Comment (XSlayer300) [inline code/modules/xenomorph/giant_worm.dm:239]:
```suggestion
		to_chat(target, "<span class='warning'>The worm cannot go there.</span>")
```

**Comment by DerrickChamberlain** [comment]
I am going to cry. My cades, my beautiful cades.

**Comment by Pariah919** [comment]
This is just the same issue pools have with a new paint really. Honestly making people have to play one type of playstyle is boring and unintuitive but hey that's just my opinion, and the "comprimise" made to attempt to make this happen is just a sad attempt to push this PR through, or make cades just irrelevant.
With psydrain this also exagerrates the lose more mechanics marines have.

**Comment by SkeletalElite** [comment]
why don't we just take the FOB and MOVE IT SOMEWHERE ELSE?

**Comment by Tupinambis** [comment]
I think the idea of essentially a hole that can bypass defenses is a nifty idea but I don't think stunning marines or damaging stuff should come with that power. The ability to get around cades alone seems more than sufficient.

**Comment by BraveMole** [comment]
> I think the idea of essentially a hole that can bypass defenses is a nifty idea but I don't think stunning marines or damaging stuff should come with that power. The ability to get around cades alone seems more than sufficient.

Then you don't realise how engie are playing. Give a way to bypass defences, and they'll do defences in their defences. The only way of doing that thing is by making it destroying any kind of counterplay the engie can imagine. Because they would imagine a ton of counterplay, and xeno siege will be as shitty as before

**Comment by SplinterGP** [comment]
Marines can just go to the other side of the cade and now xenos are trapped inside FOB, that + bombs

**Comment by BraveMole** [comment]
> Marines can just go to the other side of the cade and now xenos are trapped inside FOB, that + bombs

They can take the tunnel in the other side?

**Comment by JackTheJackhammer** [comment]
Marines should get a warning and like a rough borrowing area marker. I said ROUGH. So like a room thing. This way, they cant just borrow onto the Alamo and stealth-jack immediately. There should be a limiter on where the worm is able to dig (like the Alamo landing point and the Alamo itself). Also, I don't think a 7x7 area of destruction is good, that's like an entire cadeline (note big red LZ2 cades can all be completely obliterated if it is 7x7. I would do a 3x3 with a 7x7 light explosion radius that damages but not kills the cadeline.  

Also, your fears that the marines would drop ordinance on the worm is not a worry at all as the xenos will have breached the cadeline. If you use the worm as transportation (of which they will probably not, it sounds like more of a breaching tool than anything), the xenos would worm their way INSIDE the FOB and thus, any OB would be suicide for marines.

Also, I see this weapon as a decisive strike weapon, designed to end the round. Because of this, I advocate that the worm be able to be collapsible by destroying only one side but making it immune to fire and guns (just make them the same immunities as silos)

Also, I fear the fact that xenos could do a sneaky and burrow their way to the FOB when a major marine offensive is underway and thus pulling the marines back. This would disrupt their momentum and any retreat often results in maximum marine casualties. The fact that xenos have no proposed limitation of when they can use the worm during a marine cave offensive worries me of the CHEESE. I would limit the worm to stop this kind of cheese. This is probably the biggest "issue" I can find with this. 

Im just one person expressing their opinion and I would be surprised if you made it through this. But please do consider what I am saying.

**Comment by BraveMole** [comment]
> ALL Marines should get a warning and like a rough borrowing area marker (so the pinpointer still exists but squad marines get a warning). I said ROUGH. So like a room thing. This way, they cant just borrow onto the Alamo and stealth-jack immediately. There should be a limiter on where the worm is able to dig (like the Alamo landing point and the Alamo itself). Also, I don't think a 7x7 area of destruction is good, that's like an entire cadeline (note big red LZ2 cades can all be completely obliterated if it is 7x7. I would do a 3x3 with a 7x7 light explosion radius that damages but not kills the cadeline.
> 
> Also, your fears that the marines would drop ordinance on the worm is not a worry at all as the xenos will have breached the cadeline. If you use the worm as transportation (of which they will probably not, it sounds like more of a breaching tool than anything), the xenos would worm their way INSIDE the FOB and thus, any OB would be suicide for marines.
> 
> Also, I see this weapon as a decisive strike weapon, designed to end the round. Because of this, I advocate that the worm be able to be collapsible by destroying only one side but making it immune to fire and guns (just make them the same immunities as silos)
> 
> Also, I fear the fact that xenos could do a sneaky and burrow their way to the FOB when a major marine offensive is underway and thus pulling the marines back. This would disrupt their momentum and any retreat often results in maximum marine casualties. The fact that xenos have no proposed limitation of when they can use the worm during a marine cave offensive worries me of the CHEESE. I would limit the worm to stop this kind of cheese. This is probably the biggest "issue" I can find with this.
> 
> Im just one person expressing their opinion and I would be surprised if you made it through this. But please do consider what I am saying.

Nah i read comments.

For the "killing one part kill the other", yeah why not. It's already immune to fire, bullets and explosion. 
Marines already have an alert 5 minutes before the worm emerge, and it's not hard to understand it's going into the FOB, so stealth jack should not be an issue.

If xeno are using the worm to cheese, then they lost all silo and psy points. So marines can go back to fob, wait, and then finish the fight. 

The 7 by 7 destruction zone is arguably too big, but i'm starting to understand how engineers react to things, and i'm sure they'll do cade maze everywhere in the FOB, so this is probably needed.

And yeah, this kills cadelines, but it's a gamble from xenos, so it has to have some benefits

**Comment by DerFlammenwerfer** [comment]
> 
As long as this is implemented alongside trenches, it might not be as big an issue. You'd have a chance to set up emergency defenses.

>The only way of doing that thing is by making it destroying any kind of counterplay the engie can imagine.
Counterplay is what makes this server TGMC and not CM. 7x7 is too large of a deletion. 5x5 would be sufficient to open a nice hole in FOB cades, and from there it's up to the xenos to keep engineers from closing it back up to trap them. Aside from Big Red, Magmoor, and LV, most LZs on most maps don't really have all that much open space to work with. The ones that do have lots of space around the landing pad are a double-edged sword, as it's harder to be sure where the worm is going to pop out at.

**Comment by BraveMole** [comment]
> > Also, I don't think a 7x7 area of destruction is good, that's like an entire cadeline (note big red LZ2 cades can all be completely obliterated if it is 7x7. I would do a 3x3 with a 7x7 light explosion radius that damages but not kills the cadeline.
> 
> As long as this is implemented alongside trenches, it might not be as big an issue. You'd have a chance to set up emergency defenses, but I agree that 7x7 is a little bit too big.
> 
> > The only way of doing that thing is by making it destroying any kind of counterplay the engie can imagine.
> 
> Counterplay is what makes this server TGMC and not CM. 7x7 is too large of a deletion. 5x5 would be sufficient to open a nice hole in FOB cades, and from there it's up to the xenos to keep engineers from closing it back up to trap them. Aside from Big Red, Magmoor, and LV, most LZs on most maps don't really have all that much open space to work with. The ones that do have lots of space around the landing pad are a double-edged sword, as it's harder to be sure where the worm is going to pop out at.
> 
> Also, if engineers cade maze everywhere in the FOB, it's going to be very hard to avoid boiler gas. If anything, this burrow ability would be best used when there is a fob siege ongoing to maximize the impact it will have. The ability to just delete even 5x5 of whatever's in the worm's way is a huge boon during an ongoing siege.

I can always reduce it to 5x5 later, we will see with feedback

**Comment by DerrickChamberlain** [comment]
Possibly only meleeable, ahhhhhhh, that thing will never die.

File: code/modules/xenomorph/xeno_structures.dm [hunk 7]
```diff
@@ -49,6 +51,7 @@
 	number++
 	GLOB.xeno_resin_silos += src
 	center_turf = get_step(src, NORTHEAST)
+	RegisterSignal(SSdcs, COMSIG_GLOB_DROPSHIP_HIJACKED, .proc/destroy_silently)
```
Comment (TiviPlus) [inline code/modules/xenomorph/xeno_structures.dm:54]: unrelated changes atomize this

File: code/modules/xenomorph/resin_maw.dm [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+	for(var/mob/living/carbon/xenomorph/xeno AS in GLOB.alive_xeno_list)
+		if(xeno.loc.z != loc.z)
+			return
+		if(xeno.hivenumber != associated_hive.hivenumber)
+			return
+		if(get_dist(src, xeno) > 10)
+			return
+		xeno.frenzy_new = 6 //Extremely good pheros
+		xeno.warding_new = 6
+		xeno.recovery_new = 6
```
Comment (TiviPlus) [inline code/modules/xenomorph/resin_maw.dm:131]: but why

File: code/game/objects/objs.dm [hunk 9]
```diff
@@ -151,6 +151,9 @@
 	. = ..()
 	ENABLE_BITFIELD(obj_flags, IN_USE)
 
+/obj/proc/remove_eye_control(mob/living/user)
```
Comment (TiviPlus) [inline code/game/objects/objs.dm:154]: autodoc
